packages/aws-cdk/lib/api/hotswap/cloudformation-executable-template.ts

```diff
@@ -0,0 +1,230 @@
+    this.context = {
+      'AWS::AccountId': props.account,
+      'AWS::Region': props.region,
+      'AWS::Partition': props.partition,
+      'AWS::URLSuffix': props.urlSuffix,
+      ...props.parameters,
+    };
+    this.account = props.account;
+    this.region = props.region;
+    this.partition = props.partition;
+  }
+
+  public async findPhysicalNameFor(logicalId: string): Promise<string | undefined> {
+    const stackResources = await this.stackResources.listStackResources();
+    return stackResources.find(sr => sr.LogicalResourceId === logicalId)?.PhysicalResourceId;
+  }
+
+  public async evaluateCfnExpression(cfnExpression: any): Promise<any> {
+    if (cfnExpression == null) {
+      return cfnExpression;
+    }
+
+    if (Array.isArray(cfnExpression)) {
+      return Promise.all(cfnExpression.map(expr => this.evaluateCfnExpression(expr)));
+    }
+
+    if (typeof cfnExpression === 'object') {
+      const intrinsic = this.parseIntrinsic(cfnExpression);
+      if (intrinsic) {
+        return this.evaluateIntrinsic(intrinsic);
+      } else {
+        const ret: { [key: string]: any } = {};
+        for (const [key, val] of Object.entries(cfnExpression)) {
+          ret[key] = await this.evaluateCfnExpression(val);
+        }
+        return ret;
+      }
+    }
+
+    return cfnExpression;
+  }
+
+  async 'Fn::Join'(separator: string, args: any[]): Promise<string> {
+    const evaluatedArgs = await this.evaluateCfnExpression(args);
+    return evaluatedArgs.join(separator);
+  }
+
+  async 'Fn::Split'(separator: string, args: any): Promise<string> {
+    const evaluatedArgs = await this.evaluateCfnExpression(args);
+    return evaluatedArgs.split(separator);
+  }
+
+  async 'Fn::Select'(index: number, args: any[]): Promise<string> {
+    const evaluatedArgs = await this.evaluateCfnExpression(args);
+    return evaluatedArgs[index];
+  }
+
+  async 'Ref'(logicalId: string): Promise<string> {
+    const refTarget = await this.findRefTarget(logicalId);
+    if (refTarget) {
+      return refTarget;
+    } else {
+      throw new Error(`Reference target '${logicalId}' was not found`);
+    }
+  }
+
+  async 'Fn::GetAtt'(logicalId: string, attributeName: string): Promise<string> {
+    // ToDo handle the 'logicalId.attributeName' form of Fn::GetAtt
+    const attrValue = await this.findGetAttTarget(logicalId, attributeName);
+    if (attrValue) {
+      return attrValue;
+    } else {
+      throw new Error(`Trying to evaluate Fn::GetAtt of '${logicalId}.${attributeName}' but not in context!`);
+    }
+  }
+
+  async 'Fn::Sub'(template: string, explicitPlaceholders?: { [variable: string]: string }): Promise<string> {
+    const placeholders = explicitPlaceholders
+      ? { ...this.context, ...(await this.evaluateCfnExpression(explicitPlaceholders)) }
+      : this.context;
+
+    return template.replace(/\${([^}]*)}/g, (_: string, key: string) => {
+      if (key in placeholders) {
+        return placeholders[key];
```

Comment:
I'm pretty sure this implement is wrong -- already was wrong, nothing to do with you.

Additional lookups that I believe are always valid for `Fn::Sub` are:

- `${LogicalId}` -- evaluate `LogicalId` as a `{ Ref }` if it's not in the context object
- `${LogicalId.Attr}` -- evaluate as a `{ Fn::GetAtt }`.

Reply:
Yep, done.